pkg/operator/csidriveroperator/olmremovalcontroller.go

```diff
@@ -0,0 +1,463 @@
+	subNamespace, subName, csvName, found, err := c.findSubscription(ctx)
+	if err != nil {
+		return err
+	}
+
+	if found {
+		c.olmOperatorNamespace = subNamespace
+		// Delete the subscription, but remember the namespace of the operator
+		// in CR's annotations first - just in case this controller is
+		// restarted after it deletes the Subscription.
+		if err := c.saveOperatorNamespace(subNamespace); err != nil {
+			return err
+		}
+
+		removed, err := c.deleteCSV(ctx, subNamespace, csvName)
+		if err != nil {
+			return err
+		}
+		if !removed {
+			klog.V(4).Infof("OLMOperatorRemovalController.Sync waiting for OLM CSV to disappear")
+			return c.markProgressing("Waiting for OLM CSV to be deleted")
+		}
+
+		removed, err = c.deleteCSV(ctx, subNamespace, csvName)
+		if err != nil {
+			return err
+		}
+		if !removed {
+			klog.V(4).Infof("OLMOperatorRemovalController.Sync waiting for OLM CSV to disappear")
+			return c.markProgressing("Waiting for OLM CSV to be deleted")
+		}
+
+		removed, err = c.deleteSubscription(ctx, subNamespace, subName)
+		if err != nil {
+			return err
+		}
+		if !removed {
+			klog.V(4).Infof("OLMOperatorRemovalController.Sync waiting for OLM Subscription to disappear")
+			return c.markProgressing("Waiting for OLM Subscription to be deleted")
+		}
+	}
+
+	// The subscription has been completely deleted...
+	if c.olmOperatorNamespace == "" {
+		// Load the namespace from annotation
+		if c.olmOperatorNamespace, err = c.loadOperatorNamespace(); err != nil {
+			return err
+		}
+		klog.V(4).Infof("OLMOperatorRemovalController.Sync old namespace loaded: %q", c.olmOperatorNamespace)
+		if c.olmOperatorNamespace == "" {
+			// There is no Subscription + there is no work recorded in CR annotations: we're done!
+			klog.V(4).Infof("OLMOperatorRemovalController.Sync the old driver was not installed")
+			return c.markFinished("CSI driver installed by OLM is not preset")
+		}
+	}
+
+	// 3. Wait until OLM removes the the operator deployment
+	removed, err := c.ensureOperatorDeploymentRemoved(c.olmOperatorNamespace, c.olmOptions.OLMOperatorDeploymentName)
+	if err != nil {
+		return err
+	}
+	if !removed {
+		klog.V(4).Infof("OLMOperatorRemovalController.Sync waiting for OLM to delete the operator")
+		return c.markProgressing("Waiting for OLM to delete the operator")
+	}
+
+	// 4. Delete the driver Deployment. Note that the old operator pods can be still running at this point.
+	removed, err = c.ensureDriverDeploymentRemoved(ctx, c.olmOptions.CSIDriverNamespace, c.olmOptions.CSIDriverDeploymentName)
+	if err != nil {
+		return err
+	}
+	if !removed {
+		klog.V(4).Infof("CSIDriverMigrationController.Sync waiting for the driver Deployment to disappear")
+		return c.markProgressing("Waiting for OLM driver Deployment to be deleted")
+	}
+
+	// 5. Delete the driver DaemonSet.
+	removed, err = c.ensureDriverDaemonSetRemoved(ctx, c.olmOptions.CSIDriverNamespace, c.olmOptions.CSIDriverDaemonSetName)
```

Comment:
A pod that uses any of these PVs if gets evicted during upgrade, then it will be stuck basically?

Reply:
stuck until the new driver starts, yes